README.md
# Inexistence

> This is a seedbox script focus on Chinese users, I would prefer [QuickBox Lite](https://github.com/amefs/quickbox-lite), [QuickBox](https://quickbox.io/), [swizzin](https://swizzin.ltd), [PGBlitz](https://pgblitz.com), [rtinst](https://github.com/arakasi72/rtinst) for non-Chinese users.  
> Just a SEEDBOX script, no Plex, no Emby, no NZB support.  
> And note that this README is outdated, I'm too lazy to keep it update.  

## Notes

1. 本脚本只支持 x86_64 (amd64) 架构，其他架构都不支持。ARM 用户建议使用 [QuickBox ARM](https://github.com/amefs/quickbox-arm)  
2. 本脚本只在独服和 KVM 虚拟化的 VPS 下测试，OpenVZ、Xen 等其他虚拟化架构仍可以尝试使用，但不保证没问题  
3. 本脚本目前支持 Debian 9/10, Ubuntu 16.04/18.04. *推荐使用 Debian 10 或 Ubuntu 18.04*  
4. 本文的使用说明中的图片是一两年前的，与当前脚本存在较大出入（但文字内容是及时更新的）  
5. 建议重装完系统后使用此脚本，非全新安装的情况下（比如你先跑了个其他盒子脚本再跑这个）不确定因素太多容易翻车  
6. 目前没有简单易用的卸载方法。如果你有卸载的需求，使用前请三思  

## Usage

如果你是新手，对很多选项不甚了解，直接用这个就完事了（账号密码部分替换一下）：  
```
bash <(wget --no-check-certificate -qO- https://github.com/Aniverse/inexistence/raw/master/inexistence.sh) \
-y --tweaks --bbr --rclone --no-system-upgrade --flexget --tr-deb --filebrowser \
--de 1.3.15 --rt 0.9.8 --qb 4.1.9 -u 这十二个字换成你的用户名 -p 这十个字换成你的密码
```
如果你需要自定义安装选项：  
```
bash <(wget --no-check-certificate -qO- https://github.com/Aniverse/inexistence/raw/master/inexistence.sh)
```
短命令（就是命令长度短一些，其他方面和上边那个没任何区别）  
```
bash <(wget -qO- https://git.io/abcde)
```

## Installation Guide

![引导界面](https://github.com/Aniverse/pics/raw/master/inexistence/inexistence.01.png)

检查是否以 root 权限来运行脚本，检查公网 IP 地址与系统参数  

![升级系统](https://github.com/Aniverse/pics/raw/master/inexistence/inexistence.02.1.png)
![升级系统](https://github.com/Aniverse/pics/raw/master/inexistence/inexistence.02.2.png)

支持 `Ubuntu 16.04 / 18.04`、`Debian 9 / 10` ，`Ubuntu 14.04` 和 `Debian 7/8` 可以用脚本升级，其他系统不支持  
使用 ***`-s`*** 参数可以跳过对系统是否受支持的检查，不过这种情况下脚本能不能正常工作就是另一回事了  

![系统信息](https://github.com/Aniverse/pics/raw/master/inexistence/inexistence.03.png)

显示系统信息以及注意事项  

![安装时的选项](https://github.com/Aniverse/pics/raw/master/inexistence/inexistence.04.png)

1. ***是否升级系统***  
低于 `Ubuntu 18.04`、`Debian 10` 的 LTS 系统可以选择用脚本升级系统（Ubuntu 20.04 暂不支持）  
一般来说整个升级过程应该是无交互的，应该不会碰到什么问你 Yes or No 的问题  
升级完后会直接执行重启命令，重启完后你需要再次运行脚本来完成软件的安装  


2. ***账号密码***  
**`-u <username> -p <password>`**  
你输入的账号密码会被用于各类软件以及 SSH 的登录验证  
用户名需要以字母开头，长度 4-16 位；密码需要同时包含字母和数字，长度至少 8 位  


3. ***是否更换软件源***  
**目前默认直接换源不再提问，如果不想换源，请在运行脚本的使用 `--no-source-change` 参数**  
这个选项决定是否替换 `/etc/apt/sources.list` 文件。  
其实大多数情况下无需换源；但某些盒子默认的源可能有点问题，所以我干脆做成默认都换源了  


4. ***线程数量***  
**`--mt-single`**、**`--mt-double`**、**`--mt-half`**、**`--mt-max`**  
**目前默认直接使用全部线程不再提问，如果不想使用全部线程，请在运行脚本的使用以上的参数来指定**  
编译时使用几个线程进行编译。一般来说用默认的选项，也就是全部线程都用于编译就行了  
某些 VPS 可能限制下线程数量能避免编译过程中因为内存不足翻车  


5. ***安装时是否创建 swap***  
**`--swap`**，**`--no-swap`**  
**目前默认对于内存小于 1926MB 的服务器直接启用 swap 不再询问，如不想使用 swap 请用 `--swap-no` 参数**  
一些内存不够大的 VPS 在编译安装时可能物理内存不足，使用 swap 可以解决这个问题  
实测 1C1T 1GB 内存 的 Vultr VPS 安装 Flood 不开启 swap 的话会失败，开启就没问题了  
目前对于物理内存小于 1926MB 的都默认启用 swap，如果内存大于这个值那么你根本就不会看到这个问题……  


6. ***qBittorrent***  
**`--qb 4.2.3 --qb-static`**、**`--qb 3.3.11`**、**`--qb no`**  
static 指静态编译版本，deb 指使用 efs 菊苣编译好的 deb 包来安装。这两种安装方法的最大特点是安装速度非常快  
因为 static 和 deb 安装已经很快了，因此去除了从 repo 或 ppa 安装的选项  


7. ***Deluge***  
**`--de 1.3.15_skip_hash_check`**、**`--de 1.3.9`**、**`--de no`**  
默认选项为从源码安装 1.3.15  
2.0.3 目前运行在 Python 2.7 下，且仍然有一些 PT 站不支持 2.0.3，因此不推荐使用  
此外还会安装一些实用的 Deluge 第三方插件：  
- `AutoRemovePlus` 是自动删种插件，支持 WebUI 与 GtkUI  
- `ltconfig` 是一个调整 `libtorrent-rasterbar` 参数的插件，在安装完后就启用了 `High Performance Seed` 模式  
- `Stats`、`TotalTraffic`、`Pieces`、`LabelPlus`、`YaRSS2`、`NoFolder` 都只能在 GUI 下设置，WebUI 下无法显示  
- `Stats` 和 `TotalTraffic`、`Pieces` 分别可以实现速度曲线和流量统计、区块统计  
- `LabelPlus` 是加强版的标签管理，支持自动根据 Tracker 对种子限速，刷 Frds 可用  
- `YaRSS2` 是用于 RSS 的插件  
隐藏选项 21，是可以跳过校验、全磁盘预分配的 1.3.15 版本  
**使用修改版客户端、跳过校验 存在风险，后果自负**  


8. ***rTorrent***  
**`--rt 0.9.8`**、**`--rt 0.9.3 --enable-ipv6`**、**`--rt no`**  
这部分是调用我修改的 [rtinst](https://github.com/Aniverse/rtinst) 来安装的  
注意，`Ubuntu 18.04` 和 `Debian 9/10` 因为 OpenSSL 的原因，只能使用 0.9.6 及以上的版本，更低版本无法直接安装  
- 安装 rTorrent，ruTorrent，nginx，ffmpeg，rar，h5ai 目录列表程序  
- 0.9.2-0.9.4 支持 IPv6 用的是打好补丁的版本，属于修改版客户端  
- 0.9.6 支持 IPv6 用的是 2018.01.30 的 feature-bind 分支，原生支持 IPv6  
- 设置了 Deluge、qBittorrent、Transmission、Flexget WebUI 的反代  
- ruTorrent 版本为来自 master 分支的 3.9 版，此外还安装了如下的第三方插件和主题  
- `club-QuickBox` `MaterialDesign` 第三方主题  
- `AutoDL-Irssi` （原版 rtinst 自带）  
- `Filemanager` 插件可以在 ruTorrent 上管理文件、右键创建压缩包、生成 mediainfo 和截图  
- `ruTorrent Mobile` 插件可以优化 ruTorrent 在手机上的显示效果（不需要的话可以手动禁用此插件）  
- `Fileshare` 插件创建有时限、可自定义密码的文件分享链接  
- `GeoIP2` 插件，代替原先的 GeoIP 插件，精确度更好，支持 IPv6 地址识别  


9. **Flood**  
**`--flood`**  
**是否安装的问题已被移除，只能使用命令行参数安装**  
Flood 是 rTorrent 的另一个 WebUI，界面更为美观，加载速度快，不过功能上不如 ruTorrent  
第一次登陆时需要填写信息，端口号是 5000，挂载点是 127.0.0.1  


10. ***Transmission***  
**`--tr-deb`**、**`--tr 2.83`**、**`--tr no`**  
Transmission 默认选择从预先编译好的 deb 安装最新版 2.94（解决了文件打开数问题）  
此外还会安装 [加强版的 WebUI](https://github.com/ronggang/transmission-web-control)，更方便易用  
***隐藏和从 repo/ppa 安装的选项均已移除***  


11. ***FlexGet***  
**`--flexget`**、**`--no-flexget`**  
Flexget 是一个非常强大的自动化工具，功能非常多。大多数国内盒子用户主要用它来 RSS（它能做的事情远不止 RSS）  
目前脚本里安装 Flexget 时版本会指定为 3.0.31，同时如果系统自带的 Python3 版本低于 3.6 还会升级 Python3  
我启用了 daemon 模式和 WebUI，还预设了一些模板，仅供参考  
注意：脚本里没有启用 schedules 或 crontab，需要的话自己设置  


12. ***FileBrowser Enhanced***  
**`--filebrowser`**、**`--no-filebrowser`**  
File Browser 提供了网页文件管理器的功能, 可以用于上传、删除、预览、重命名以及编辑盒子上的文件  
脚本安装的是 [荒野无灯的 Docker 版 FileBrowser Enhanced](https://hub.docker.com/r/80x86/filebrowser)，[功能更加强大](https://raw.githubusercontent.com/ttys3/filebrowser-enhanced/master/FBvsFBE.zh.png)  
这个增强版还可以在网页上右键获取文件的 mediainfo、制作种子、截图、解压等等，对 PT 来说也非常实用  
还有一个在 http://ip:7576 网址、使用 root 运行、挂载 / 目录的 filebrowser，需要输入 `systemctl enable filebrowser-root --now` 手动启用  


14. ***系统设置***  
**`--tweaks`**、**`--no-tweaks`**  
默认启用，具体操作如下：  
- 安装 [vnstat](https://github.com/vergoh/vnstat) 2.6 以及 [vnstat dashboard](https://github.com/alexandermarston/vnstat-dashboard/)，可以在网页上查看流量统计  
- （注：vnstat dashboard 使用的前提是用脚本安装了 rTorrent，且在 Debian 8 下不可用）  
- 修改时区为 UTC+8  
- 语言编码设置为 en.UTF-8  
- 设置 `alias` 简化命令（私货夹带）  
- 修改 screenrc 设置  
- 将最大可用空间的硬盘分区的 Linux 保留空间调整到 1%（原先是 5%）  


15. ***Remote Desktop***  
**`--vnc`**、**`--x2go`**  
**是否安装的问题已被移除，只能使用命令行参数安装**  
远程桌面可以完成一些 CLI 下做不了或者 CLI 实现起来很麻烦的操作，比如 BD-Remux，wine uTorrent  
推荐使用 noVNC，网页上即可操作  


16. ***wine / mono***  
**`--wine`、`--mono`**  
**是否安装的问题已被移除，只能使用命令行参数安装**  
`wine` 可以实现在 Linux 上运行 Windows 程序，比如 DVDFab、uTorrent  
`mono` 是一个跨平台的 .NET 运行环境，BDinfoCLI、Jackett、Sonarr 等软件的运行都需要 mono   


17. ***rclone***  
**`--rclone`**  
**是否安装的问题已被移除，只能使用命令行参数安装**  
rclone 是一个强大的网盘同步工具。默认不安装。安装好后需要自己输入 rclone config 进行配置  
此外这个选项还会安装 [gclone](https://github.com/donwa/gclone)  


18. ***Some additional tools***  
**`--tools`**  
**是否安装的问题已被移除，只能使用命令行参数安装**  
安装下列软件：  
- `mediainfo` 用最新版是因为某些站发种填信息时有这方面的要求，比如 HDBits  
- `mkvtoolnix` 主要是用于做 BD-Remux  
- `eac3to` 需要 wine 来运行，做 remux 时用得上  
- `ffmpeg` 对于大多数盒子用户来说主要是拿来做视频截图用，安装的是静态编译版本  


19. ***BBR***  
**`--bbr`**、**`--no-bbr`**  
**是否安装的问题已被移除，只能使用命令行参数安装**  
（如果你想安装魔改版 BBR 或 锐速，请移步到 [TrCtrlProToc0l](https://github.com/Aniverse/TrCtrlProToc0l) 脚本）  
会检测你当前的内核版本，大于 4.9 是默认不安装新内核与 BBR，高于 4.9 是默认直接启用BBR（不安装新内核）  
注意：更换内核有风险，可能会导致无法正常启动系统  


20. ***libtorrent-rasterbar***  
**`--lt RC_1_1`**、**`--lt RC_1_0`**、**`--lt system`**、**`--lt 1.1.12`**  
**选择哪个版本的问题已被移除，默认使用 RC_1_1，只能使用命令行参数自行指定**  
libtorrent-rasterbar 是 Deluge 和 qBittorrent 所使用的后端，除非 qBittorrent 使用静态编译版本，不然只要选择安装 Deluge 和 qBittorrent 中的任意一样，libtorrent 都是必装的。鉴于 lt 与 de/qb 兼容的情况比较复杂，现在脚本里直接统一使用 libtorrent RC_1_1（版本号 1.1.14）。如果你需要自定义版本号，请使用 `--lt <version>` 参数（自定义版本时，不保证脚本能正常工作）  



![确认信息](https://github.com/Aniverse/pics/raw/master/inexistence/inexistence.05.png)

如果你哪里写错了，先退出脚本重新选择；没什么问题的话就敲回车继续  
使用 ***`-y`*** 可以跳过开头的信息确认和此处的信息确认，配合其他参数可以做到无交互安装  

![使用参数](https://github.com/Aniverse/pics/raw/master/inexistence/inexistence.10.png)



-------------------



![安装完成界面](https://github.com/Aniverse/pics/raw/master/inexistence/inexistence.06.png)

安装完成后会输出各类 WebUI 的网址，以及本次安装花了多少时间，然后问你是否重启系统（默认是不重启）  

![安装失败界面](https://github.com/Aniverse/pics/raw/master/inexistence/inexistence.07.png)

如果报道上出现了偏差，会提示你如何查看日志（报错时请务必附上日志！）  

![WebUI](https://github.com/Aniverse/pics/raw/master/inexistence/inexistence.08.png)

最后打开浏览器检查下各客户端是否都能正常访问，一般是没问题的……  





## mingling

**这个脚本我基本放弃治疗的，很少更新，会有过时的问题**  
不做具体的介绍了，直接看图吧  


ChangeLOG.md
# To do list

近期  

* [x] inexistence.sh 启用 --help，解释各参数的作用与用法  
* [x] IPv6 脚本更新  
* [x] 启用 flexget/install  
* [x] 启用 flexget/configure  
* [x] 启用 vnstat/install  
* [x] 启用 transmission/configure  
* [x] 启用 novnc  
* [x] 启用 filebrowser  
* [x] 启用 deluge/configure  
* [ ] 启用 deluge/install，只支持安装 1.3  
* [ ] deluge/install 支持 2.0.3 的 efs deb  
* [x] 清理项目中的不再使用的大文件，缩减项目体积（主要是 deluge plugins）  
* [ ] 清理完以后发布 release，主要是备份和方便以后跨版本对比  
* [ ] 升级 bdinfocli 的版本  
* [x] 各软件的日志链接到一个目录内，方便在 h5ai／FileBrowser 中查看  
* [x] 各软件的配置文件链接到一个目录内，方便在 h5ai／FileBrowser 中查看  

中期  

* [x] 编写 deb-get，从 efs 的软件仓库中下载 deb，并设有 fallback  
* [x] 放弃 Debian 8 支持  
* [x] 移除脚本里的换源、tools、bbr、wine 问题  
* [x] 移除脚本里的 install from ppa. repo  
* [x] 移除 libtorrent 1.0 选项  
* [x] 启用 qbittorrent/install  
* [ ] 启用 flood/install  
* [ ] 启用 flood/configure  
* [ ] 启用 x2go/install  
* [ ] 启用 art/configure  
* [ ] 启用 transmission/install  
* [ ] 支持 Transmission 3.0  
* [x] 新 App：autoremove-torrents  
* [ ] 新 App：Jackett  
* [ ] 新 App：The Lounge  
* [ ] deluge/install 支持 2.0.3 的 building  
* [ ] 支持 python3-libtorrent building  
* [ ] 支持 libtorrent 1.2.x  
* [ ] `hezi` 脚本支持 `hezi install qb 4.2.5` 这样的用法  
* [ ] 补全 changelog，即补完我在使用 git 之前的更新记录  
* [ ] 大幅更新 README  
* [x] 移除升级系统的功能，要升级系统请用 [DieNacht 的脚本](https://github.com/DieNacht/debian-ubuntu-upgrade)  

远期  

* [ ] 支持 Ubuntu 20.04 LTS  
* [x] 使用 pyenv  
* [ ] 使主脚本 inexistence.sh 有完整的中文交互界面  
* [ ] 不使用 rtinst，自己写 rTorrent 与 ruTorrent 的安装  
* [ ] 启用 ftp/install  
* [ ] 启用 nginx/configure  
* [ ] `hezi` 脚本里加入新增用户、删除用户、改用户密码的功能  








# ChangeLog  





## [2020.07.26/08.05/08.09/08.17, 6 commits](https://github.com/Aniverse/inexistence/compare/620e4d3...bd189a8)  

`the inexistence project`  
1. Refactor：[BitTorrentClientCollection](https://github.com/Aniverse/BitTorrentClientCollection) 文件重新上传  
- 清空历史提交记录  
- 移除 deluge.plugins、firmware、miscellaneous，这些都放到了 [inexistence-files](https://github.com/Aniverse/inexistence-files) 内  
- 移除 tcp.cc，直接从 qss 的 repo 里下载  
- 新增 4.20-5.7 的 liquorix 内核（sid／buster 版）  
2. Refactor：[inexistence-files](https://github.com/Aniverse/inexistence-files) 文件重新上传  
- 清空历史提交记录  
- 移除 deb 文件，那些文件可以从 [quickbox-files](https://github.com/amefs/quickbox-files) 下载  
- 加入 ioping、fio、iperf3、smartctl 的静态编译版本  
- 加入一些 patch 文件，如 de 跳校验、老版 rt 支持 ipv6、qbt.webui.table 等  
- 更新 README  
3. Refactor：XanMod 内核的下载移动到 [XanMod-DL](https://github.com/Aniverse/XanMod-DL)  
- 目前有 4.9-5.7 的 XanMod 内核可供下载  

`inexistence 1.2.7.9`  
1. Feature：移除 speedtest-cli 的安装（避免与官方版本冲突）  
2. BugFix：Transmission 的安装使用 deb-get  

`deluge/configure r10011`  
1. Feature：移除 reannounce-0.1-py3.7.egg  

`filebrowser r11014`  
1. Feature：使用最新版 filebrowser release  

`s-alias r12022`  
1. Feature：jieya 支持解压 tar.xz 文件了，并可以选择删除源文件  
2. NewFunction：chaip，使用 ipapi 查 ip  
3. NewFunction：s-debug  
4. BugFix：修复注释了 setcolor 后，HISTTIMEFORMAT 无法显示高亮色彩的问题  





## [2020.07.18-20, 10 commits](https://github.com/Aniverse/inexistence/compare/ee983e3...620e4d3)  

`the inexistence project`  
1. **Codes：彻底删除了项目中的 deluge 插件（commits 历史中也删除了），减少了整个项目的体积**  

`inexistence 1.2.7.7`  
1. UI：系统可以升级时，加入 DieNacht 的升级系统脚本的提示信息  
2. UI：只有当 `lt_version` 存在且不为 RC_1_1 的时候才显示所选择的版本信息  
3. UI：`END_output_url` 输出时去掉结尾多余的空格  

`function r14140`  
1. **BugFix：配置文件和日志的链接从硬链改成了相对路径的软链**  
2. BugFix：修复 END_output_url 中 transmission 链接不包含密码的问题  
3. BugFix：修复 END_output_url 中 rtorrent 和 flood 的失败没有正确输出日志  
4. UI：rtorrent 失败时的两个日志合并上传到 sprunge.us  
5. set_log_when_there_is_none：不再 echo > $HOME/.123.log 的内容  

`s-alias r12019`
1. s-info：支持显示 CPU、内存、硬盘信息，增加网络配置和软件源配置文件的路径显示  

`ask r11041`  
1. BugFix：`create_log_link` 之前先 `export_inexistence_info`  

`xiansu r20003`  
1. BugFix：改进网卡的判断方式  
2. Feature：默认限速速度改为 50Mbps  





## [2020.07.07/15, 4+30 commits](https://github.com/Aniverse/inexistence/compare/5d302b6...3ff4800)  

`the inexistence project`  
1. **Codes：删除系统升级相关的代码**  
2. **NewFeature：将各软件的日志和配置文件链接到一个目录内，方便查看**  
3. **NewFeature：使用 deb-get 下载 QuickBox-Lite 的 deb**  
默认从 GitHub 下载，失败的话则依次从 CF、OSDN、SourceForge 下载，全都失败的话再提示  
4. NewOpts：`--qb-source`  
这个选项强制让 qb 从源码编译安装  
5. ClientScript：`export_inexistence_info` 后做 `check_var_iUser_iPass_iHome` 的检查  

Config/mktorrent.php
<?php
/*
MkTorrent WebUI
Copyright (C) 2009 RustyBadger

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/gpl-3.0.txt>.

The author of this software can be contacted at rustybadger at users.sourceforge.net.
*/

// Config
$root = "/home/REPLACEUSERNAME"; // Set this to the directory you want to expose (e.g. /home/rtorrent)
$mktorrent = "/usr/bin/mktorrent"; // This should be (for security reasons) the full path to mktorrent. (It can be found by running `which mktorrent` if you're unsure.)

function normalise($path) { // This normalises a path to its simplest form, and also prevents jailbreaking.
	$normalised = array();
	foreach (explode("/", $path) as $p) {
		if ($p == "..") {array_pop($normalised);}
		else if ($p == ".") {}
		else if ($p) {array_push($normalised, $p);}
	}
	array_unshift($normalised, ""); // This means the path always starts with /
	return implode("/", $normalised);
}

$_POST['input'] = normalise($_POST['input']);
$_POST['output'] = normalise($_POST['output']);

if ($_POST['inputnav'] != "" && isset($_POST['input']) && is_dir($root . $_POST['input'])) {$_POST['inputpath'] = $_POST['input'];}
if ($_POST['outputnav'] != "" && isset($_POST['output']) && is_dir($root . $_POST['output'])) {$_POST['outputpath'] = $_POST['output'];}

$command = $mktorrent . " ";
if (is_array($_POST['announce'])) {foreach ($_POST['announce'] as $a) {if ($a != "") {$command .= "-a " . escapeshellarg($a) . " ";}}} else {$_POST['announce'] = array();}
if ($_POST['public'] != "") {if ($_POST['public'] == "false") {$command .= "-p ";}} else {$_POST['public'] = "true";}
if ($_POST['piecelength'] != "") {$command .= "-l " . escapeshellarg($_POST['piecelength']) . " ";}
if (is_array($_POST['webseeds'])) {foreach ($_POST['webseeds'] as $w) {if ($w != "") {$command .= "-w " . escapeshellarg($w) . " ";}}} else {$_POST['webseeds'] = array();}
if ($_POST['name'] != "") {$command .= "-n " . escapeshellarg($_POST['name']) . " ";}
if ($_POST['comment'] != "") {$command .= "-c " . escapeshellarg($_POST['comment']) . " ";}
if ($_POST['creation'] != "") {if ($_POST['creation'] == "false") {$command .= "-d ";}} else {$_POST['creation'] = "true";}
if ($_POST['verbose'] != "") {if ($_POST['verbose'] == "false") {$command .= "-v ";}} else {$_POST['verbose'] = "true";}
if ($_POST['outputfile'] != "") {$command .= "-o " . escapeshellarg($root . $_POST['outputpath'] . "/" . $_POST['outputfile']) . " ";} else {$_POST['outputfile'] = ".torrent";}
if ($_POST['input'] != "") {$command .= escapeshellarg($root . $_POST['input']);}

if ($_POST['report'] == "") {$_POST['report'] = "false";}
if ($_POST['advanced'] == "") {$_POST['advanced'] = "false";}

if ($_POST['inputnav'] == "" && $_POST['outputnav'] == "") {exec($command, $returnstr, $return);}
else {$command = ""; $returnstr = array(); $return = -1;}
?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8" />
<title>MkTorrent WebUI</title>
<script type="text/javascript">
function showhide(id, input) {
	if (document.getElementById(id).style.display == "none") {
		document.getElementById(id).style.display = "";
		document.getElementById(input).value = "true";
	} else {
		document.getElementById(id).style.display = "none";
		document.getElementById(input).value = "false";
	}
}
function label(id, check) {document.getElementById(id).focus(); if (check) {document.getElementById(id).checked = true;}}<?php // This is because a label outside of a td doesn't work, and inside only the text is clickable. ?>
function insertRow(table, name) {
	var row = table.insertRow(table.rows.length - 1);
	var cell = document.createElement("td"); row.appendChild(cell);
	var input = document.createElement("input");
		input.type = "text";
		input.name = name;
		input.onblur =  function() {if (this.value == "") {deleteRow(table, row.rowIndex);}};
	cell.appendChild(input);
		input.focus();
}
function deleteRow(table, index) {
	table.deleteRow(index);
}
</script>
<style type="text/css">
h2 {text-decoration: underline;}
table, select, input {width: 100%;}
input[type=radio] {width: auto;}
</style>
</head>
<body>
<form action="<?php echo $_SERVER['REQUEST_URI']; ?>" method="post">
<h2><a href="javascript: showhide('reportid', 'report');">Report</a><input type="hidden" name="report" id="report" value="<?php echo $_POST['report']; ?>" /></h2>
<p><?php switch ($return) {case -1: echo ""; break; case 0: echo "MkTorrent Ran Sucessfully."; break; default: echo "MkTorrent Reported An Error."; break;} ?></p>
<pre id="reportid"<?php echo ($_POST['report'] == "true" ? "" : " style=\"display: none;\""); ?>>
<?php
if ($command != "") {echo "# " . htmlspecialchars(str_replace(array($root, $mktorrent), array("", "mktorrent "), $command)) . "\n";}
echo htmlspecialchars(str_replace($root, "", implode("\n", $returnstr)));
?>
</pre>
<h2>Options</h2>
<table>
<tr><td onclick="label('announce');">Tracker Groups:</td><td>
<table id="announcetable">
<?php
foreach ($_POST['announce'] as $i) {if ($i != "") {echo "<tr><td><input type=\"text\" name=\"announce[]\" value=\"" . htmlspecialchars($i) . "\" onblur=\"if (this.value == '') {deleteRow(document.getElementById('announcetable'), this.parentNode.parentNode.rowIndex);}\" /></td></tr>\n";}}
?>
<tr><td><input type="text" name="announce[]" id="announce" onfocus="insertRow(document.getElementById('announcetable'), 'announce[]');" /></td></tr>
</table>
</td></tr>
<tr><td rowspan="2">Public Torrent:</td><td onclick="label('publictrue', true);"><input type="radio" name="public" id="publictrue" value="true"<?php echo ($_POST['public'] == "true" ? " checked=\"checked\"" : ""); ?> /> Public</td></tr>
<tr><td onclick="label('publicfalse', true);"><input type="radio" name="public" id="publicfalse" value="false"<?php echo ($_POST['public'] == "false" ? " checked=\"checked\"" : ""); ?> /> Private</td></tr>
</table>
<table>
<tr><td style="width: 50%;" onclick="label('input');">Input File/Directory:</td><td colspan="2" style="width: 50%;" onclick="label('outputfile');">Output Torrent:</td></tr>
<tr><td rowspan="2">
<input type="hidden" name="inputpath" value="<?php echo htmlspecialchars($_POST['inputpath']); ?>" />
<input type="hidden" name="inputnav" />
<select name="input" id="input" size="16" ondblclick="this.form.inputnav.value = 'true'; this.form.submit();" onkeypress="if (event.keyCode == 13) {this.form.inputnav.value = 'true'; this.form.submit();};">
<?php
if ($dir = dir($root . $_POST['inputpath'])) {
	$nodes = array();
	while ($node = $dir->read()) {array_push($nodes, $node);} $dir->close();
	natcasesort($nodes);
	foreach ($nodes as $node) {echo "<option value=\"" . htmlspecialchars($_POST['inputpath'] . "/" . $node) . "\"" . ($_POST['input'] == $_POST['inputpath'] . "/" . $node ? " selected=\"selected\"" : "") . ">" . htmlspecialchars($node) . "</option>\n";}
} else {echo "<option value=\"" . htmlspecialchars($_POST['inputpath'] . "/.") . "\">.</option>\n<option value=\"" . htmlspecialchars($_POST['inputpath'] . "/..") . "\">..</option>\n<option value=\"" . htmlspecialchars($_POST['inputpath'] . "/") . "\">Failed to Open Directory</option>\n";}
?>
</select>
</td><td><input type="text" name="outputfile" id="outputfile" value="<?php echo htmlspecialchars($_POST['outputfile']); ?>" /></td><td>in</td></tr>
<tr><td colspan="2">
<input type="hidden" name="outputpath" value="<?php echo htmlspecialchars($_POST['outputpath']); ?>" />
<input type="hidden" name="outputnav" />
<select name="output" size="14" ondblclick="this.form.outputnav.value = 'true'; this.form.submit();" onkeypress="if (event.keyCode == 13) {this.form.outputnav.value = 'true'; this.form.submit();};" onblur="this.options[this.selectedIndex].selected = false;">
<?php
if ($dir = dir($root . $_POST['outputpath'])) {
	$nodes = array();
	while ($node = $dir->read()) {array_push($nodes, $node);}
	$dir->close();
	natcasesort($nodes);
	foreach ($nodes as $node) {echo "<option value=\"" . htmlspecialchars($_POST['outputpath'] . "/" . $node) . "\">" . htmlspecialchars($node) . "</option>\n";}
} else {echo "<option value=\"" . htmlspecialchars($_POST['outputpath'] . "/.") . "\">.</option>\n<option value=\"" . htmlspecialchars($_POST['outputpath'] . "/..") . "\">..</option>\n<option value=\"" . htmlspecialchars($_POST['outputpath'] . "/") . "\">Failed to Open Directory</option>\n";}
?>
</select>
</td></tr>
</table>
<h2><a href="javascript: showhide('advancedid', 'advanced');">Advanced Options</a><input type="hidden" name="advanced" id="advanced" value="<?php echo $_POST['advanced']; ?>" /></h2>
<table id="advancedid"<?php echo ($_POST['advanced'] == "true" ? "" : " style=\"display: none;\""); ?>>
<tr><td onclick="label('piecelength');">Piece Length (in 2^n bytes):</td><td><input type="text" name="piecelength" id="piecelength" value="<?php echo htmlspecialchars($_POST['piecelength']); ?>" /></td></tr>
<tr><td onclick="label('webseeds');">Web Seeds URLs:</td><td>
<table id="webseedstable">
<?php
foreach ($_POST['webseeds'] as $i) {if ($i != "") {echo "<tr><td><input type=\"text\" name=\"webseeds[]\" value=\"" . htmlspecialchars($i) . "\" onblur=\"if (this.value == '') {deleteRow(document.getElementById('webseedstable'), this.parentNode.parentNode.rowIndex);}\" /></td></tr>\n";}}
?>
<tr><td><input type="text" name="webseeds[]" id="webseeds" onfocus="insertRow(document.getElementById('webseedstable'), 'webseeds[]');" /></td></tr>
</table>
</td></tr>
<tr><td onclick="label('name');">Torrent Name:</td><td><input type="text" name="name" id="name" value="<?php echo htmlspecialchars($_POST['name']); ?>" /></td></tr>
<tr><td onclick="label('comment');">Comment:</td><td><input type="text" name="comment" id="comment" value="<?php echo htmlspecialchars($_POST['comment']); ?>" /></td></tr>
<tr><td rowspan="2">Write Creation Date:</td><td onclick="label('creationtrue', true);"><input type="radio" name="creation" id="creationtrue" value="true"<?php echo ($_POST['creation'] == "true" ? " checked=\"checked\"" : ""); ?> /> Yes</td></tr>
<tr><td onclick="label('creationfalse', true);"><input type="radio" name="creation" id="creationfalse" value="false"<?php echo ($_POST['creation'] == "false" ? " checked=\"checked\"" : ""); ?> /> No</td></tr>
<tr><td rowspan="2">Be Verbose:</td><td onclick="label('verbosetrue', true);"><input type="radio" name="verbose" id="verbosetrue" value="true"<?php echo ($_POST['verbose'] == "true" ? " checked=\"checked\"" : ""); ?> /> Yes</td></tr>
<tr><td onclick="label('verbosefalse', true);"><input type="radio" name="verbose" id="verbosefalse" value="false"<?php echo ($_POST['verbose'] == "false" ? " checked=\"checked\"" : ""); ?> /> No</td></tr>
</table>
<div><button type="submit">MkTorrent!</button></div>
</form>
<p><a href="http://sourceforge.net/projects/mktorrentwebui" rel="external" onclick="window.open(this.href); return false;"><img src="http://sflogo.sourceforge.net/sflogo.php?group_id=260083&amp;type=10" width="80" height="15" alt="Check out our SourceForge project page" /></a></p>
</body>
</html>
